File: tests/data_structures/hash_table/hash_table.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func randomString(length int) string {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	result := make([]byte, length)
	for i := range result {
		result[i] = charset[rand.Intn(len(charset))]
	}
	return string(result)
}

func main() {
	rand.Seed(time.Now().UnixNano())
	numOperations := 100000
	
	// Generate test data
	keys := make([]string, numOperations)
	values := make([]int, numOperations)
	
	for i := 0; i < numOperations; i++ {
		keys[i] = randomString(10)
		values[i] = rand.Intn(1000000) + 1
	}
	
	fmt.Printf("Testing hash table with %d operations...\n", numOperations)
	totalStart := time.Now()
	
	// Create hash table (map in Go)
	hashTable := make(map[string]int)
	
	// Insert operations
	insertStart := time.Now()
	for i := 0; i < numOperations; i++ {
		hashTable[keys[i]] = values[i]
	}
	insertTime := time.Since(insertStart)
	
	// Lookup operations
	lookupStart := time.Now()
	foundCount := 0
	for _, key := range keys {
		if _, exists := hashTable[key]; exists {
			foundCount++
		}
	}
	lookupTime := time.Since(lookupStart)
	
	// Delete operations (every other key)
	deleteStart := time.Now()
	deletedCount := 0
	for i := 0; i < numOperations; i += 2 {
		if _, exists := hashTable[keys[i]]; exists {
			delete(hashTable, keys[i])
			deletedCount++
		}
	}
	deleteTime := time.Since(deleteStart)
	
	totalTime := time.Since(totalStart)
	
	fmt.Println("Result:")
	fmt.Printf("  Inserted: %d items\n", numOperations)
	fmt.Printf("  Found: %d/%d items\n", foundCount, numOperations)
	fmt.Printf("  Deleted: %d items\n", deletedCount)
	fmt.Printf("  Remaining: %d items\n", len(hashTable))
	fmt.Println("Timing:")
	fmt.Printf("  Insert time: %.6f seconds\n", insertTime.Seconds())
	fmt.Printf("  Lookup time: %.6f seconds\n", lookupTime.Seconds())
	fmt.Printf("  Delete time: %.6f seconds\n", deleteTime.Seconds())
	fmt.Printf("  Total time: %.6f seconds\n", totalTime.Seconds())
}

File: tests/data_structures/hash_table/hash_table.ts
#!/usr/bin/env node

function randomString(length: number = 10): string {
    const charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
        result += charset.charAt(Math.floor(Math.random() * charset.length));
    }
    return result;
}

function main(): void {
    const numOperations = 100000;
    
    // Generate test data
    const keys: string[] = [];
    const values: number[] = [];
    
    for (let i = 0; i < numOperations; i++) {
        keys.push(randomString(10));
        values.push(Math.floor(Math.random() * 1000000) + 1);
    }
    
    console.log(`Testing hash table with ${numOperations} operations...`);
    const totalStart = process.hrtime.bigint();
    
    // Create hash table (Map in TypeScript)
    const hashTable = new Map<string, number>();
    
    // Insert operations
    const insertStart = process.hrtime.bigint();
    for (let i = 0; i < numOperations; i++) {
        hashTable.set(keys[i], values[i]);
    }
    const insertTime = Number(process.hrtime.bigint() - insertStart) / 1e9;
    
    // Lookup operations
    const lookupStart = process.hrtime.bigint();
    let foundCount = 0;
    for (const key of keys) {
        if (hashTable.has(key)) {
            foundCount++;
        }
    }
    const lookupTime = Number(process.hrtime.bigint() - lookupStart) / 1e9;
    
    // Delete operations (every other key)
    const deleteStart = process.hrtime.bigint();
    let deletedCount = 0;
    for (let i = 0; i < numOperations; i += 2) {
        if (hashTable.has(keys[i])) {
            hashTable.delete(keys[i]);
            deletedCount++;
        }
    }
    const deleteTime = Number(process.hrtime.bigint() - deleteStart) / 1e9;
    
    const totalTime = Number(process.hrtime.bigint() - totalStart) / 1e9;
    
    console.log("Result:");
    console.log(`  Inserted: ${numOperations} items`);
    console.log(`  Found: ${foundCount}/${numOperations} items`);
    console.log(`  Deleted: ${deletedCount} items`);
    console.log(`  Remaining: ${hashTable.size} items`);
    console.log("Timing:");
    console.log(`  Insert time: ${insertTime.toFixed(6)} seconds`);
    console.log(`  Lookup time: ${lookupTime.toFixed(6)} seconds`);
    console.log(`  Delete time: ${deleteTime.toFixed(6)} seconds`);
    console.log(`  Total time: ${totalTime.toFixed(6)} seconds`);
}

main();

